src/task.rs

```diff
@@ -130,16 +145,111 @@ impl OptimizePlanTask {
+}
+
+impl Clone for EnforceAndCostTask {
+    fn clone(&self) -> EnforceAndCostTask {
+        todo!()
+    }
 }
 
 impl EnforceAndCostTask {
-    pub const fn new(plan: GroupPlanRef) -> Self {
-        EnforceAndCostTask { _plan: plan }
+    pub const fn new(new_plan: GroupPlanRef, new_required_prop: Rc<PhysicalProperties>) -> Self {
+        EnforceAndCostTask {
+            plan: new_plan,
+            required_prop: new_required_prop,
+            prev_index: 0,
+        }
     }
 
-    fn execute(self, _task_runner: &mut TaskRunner, _optimizer_ctx: &mut OptimizerContext) {
+    fn make_required_props_list(&self) -> Vec<Vec<PhysicalProperties>> {
+        self.plan.borrow().operator().get_reqd_prop()
+    }
+
+    fn add_enforcer_to_group(&self, _output_prop: &PhysicalProperties) -> &GroupPlanRef {
+        // add enforcer according to different specifications
         todo!()
+        // let plan = output_prop.order_property.add_enforcer();
+    }
+
+    fn prev_index(&self) -> usize {
+        self.prev_index
+    }
+
+    fn submit_cost_plan(&self, child_output_props: &[Rc<PhysicalProperties>]) {
+        let mut curr_plan = &self.plan;
+        let output_prop = curr_plan.borrow().operator().derive_output_prop(child_output_props);
+
+        let mut curr_cost = curr_plan.borrow().compute_cost();
+        if !output_prop.satisfy(&self.required_prop) {
+            // add enforcer
+            curr_plan = self.add_enforcer_to_group(&output_prop);
+            curr_cost = curr_plan.borrow().compute_cost();
+        }
+        match curr_plan
+            .borrow()
+            .group()
+            .borrow()
+            .lowest_cost_plans()
+            .get(&self.required_prop)
+        {
+            None => {}
+            Some((cost, _group_plan)) => {
+                // if current cost is larger, do nothing
+                if curr_cost.cost_value() >= cost.cost_value() {
+                    return;
+                }
+            }
+        }
+        // update lowest_cost_plans
+        curr_plan
+            .borrow_mut()
+            .group()
+            .borrow_mut()
+            .lowest_cost_plans_mut()
+            .insert(self.required_prop.clone(), (curr_cost, (*curr_plan).clone()));
+    }
+
+    /**
+     * 1. make require property for children base of current operator
+     * 2. try to optimize child group and get best (Cost, GroupPlan) pair of every children
+     * 3. once we get all output property of one candidate loop, derive output property base of current operator
+     * 4. if output property does not satisfied require property, add enforcers and submit (Cost, GroupPlan) pair
+     */
+    fn execute(mut self, task_runner: &mut TaskRunner, _optimizer_ctx: &mut OptimizerContext) {
+        // 1. according to current operator create new requestPropList for children
+        let reqd_props_list = self.make_required_props_list();
+        for (index, required_props) in reqd_props_list.iter().enumerate() {
```

Comment:
1. for (index, required_props) in reqd_props_list.iter().enumerate().skip(self.prev_index())
2. then delete 
```
if index < self.prev_index() {
    continue;
}
```

Reply:
get